text/3371-cargo-target-base-dir.md

```diff
@@ -0,0 +1,409 @@
+
+In the same vein, `cargo metadata` fills the target directory information with the absolute path and make no mention of `CARGO_TARGET_BASE_DIR` since it can only be used with a single workspace at once.
+
+### `cargo clean`
+
+Currently, if `CARGO_TARGET_DIR` is set to anything but `target` for a project, `cargo clean` does not delete the `target/` directory if it exists, instead deleting the directory pointed by `CARGO_TARGET_DIR`. The same behavior is used for `CARGO_TARGET_BASE_DIR`: if it set, `cargo clean` deletes `CARGO_TARGET_BASE_DIR/<project-path>/` and not `target/`.
+
+### Providing forward links
+
+[`targo`][tg] provides forward link (it links from `<workspace>/target` to its own target directory) as a way for existing tools to continue working despite there being no `CARGO_TARGET_DIR` set for them to find the real target directory.
+
+`cargo` does not provide them for `CARGO_TARGET_DIR`. This is not a limitation when using the environment variable set globally, since all processes can read it, but it is one when this config is only set on specific calls or via `target-dir` in the config, meaning others tools cannot easily pick it up (and most external tools don't use `cargo-metadata`, which makes them all broken by default, but fixing this situation is not this RFC's purpose).
+
+When `CARGO_TARGET_BASE_DIR` is used (in any form) and not superseded by other configurations (`CARGO_TARGET_DIR`), it *will* use a forward link by adding a `target` symlink to the real target directory. This `target` symlink will be in the exact place the real target directory would have been if `CARGO_TARGET_BASE_DIR` and `CARGO_TARGET_DIR` weren't set at all.
+
+This has a two big advantages: not breaking external tools and giving easy access to artifacts produced by `cargo build/test/doc` to users (they're in the habit of typing `./target/debug/my-bin`, this would continue working with forward links).
+
+A config option (CLI, `config.toml` and env var), `link-target-dir`, will be introduced to deactivate this behaviour, but it will `true` by default, for the reasons provided in favor of forward links just above.
+
+#### Detailed working of forward links
+
+When creating a forward link `cargo` will first attempt to create a symbolic link (regardless of the platform). If that fails, it will attempt zero or more platform-specific solutions, like junction points on NTFS. If that fails too, a warning will be emitted but this will not prevent the rest of the action to go on: regular calls like `cargo check/clippy/build/test` likely won't need this forward link and after the user has been warned they could either resolve the problem themselves or ignore it, depending on their own use case and domain-specific knowledge.
```

Comment:
From the motivation

> Allows easily having separate directories for Rust-Analyzer and Cargo itself

What happens if a `target/` dir or symlink already exists?  This is important both for transition and for the rust-analyzer use case.

Reply:
1. We do nothing: low initial surprise, potential for bad behaviour later when the user tries to run binaries that are out of date
  - We could print a clear warning then to make it visible the cache is out of date ?
3. We delete and replace with a symlink: higher initial surprise, more compatibility with existing programs and tools probably ?

Reply:
Having target forward link is a feature commonly used in package build systems like FreeBSD ports, MacPorts, etc.

Reply:
If we don't update the symlink, then if someone updates their coonfig, it will be stale

If we do update it, it will break workflows where people set a per-tool target dir (CLI, rust-analyzer, etc).

Almost seems like we'd want to specialize config file (fairly static) with env variables / `--target-dir` (more dynamic) but even that could likely cause problems and be even more confusing because nothing else in cargo acts like this.

Reply:
Usually package systems I mentioned use following: 

* whole path is used as a subfolder aster prefix, not hash. E.g. for a project in /path/to/projest build folder is created under /prefix/path/to/folder. 
* These folders and symlink are created when project is built (pre-build step, if symlink or folder doesn't exists)
* symlinks are used by build systems and all tools. This ease algorithms used in system itself, tools and make behaviour backward compatible if anything changes. 
* symlinks and build folders are removed on clean
* if a user just removed symlink and moved project folder, he's responsible for a cleanup.

Reply:
@eirnym does that prior art deal with the project being concurrently built with different target directories?

Reply:
What we likely should do is if people have background builders where they don't care about the artifacts, tell them to set `target-dir-link = `false`.

Reply:
@epage I agree, that on Windows it would be a problem, I use mostly macOS, Linux and FreeBSD, so I mentioned only best systems I've seen so far. Hashes would be great if a user can cleanup caches for an exact project without just removing everything because it's impossible to find it by just looking on it.

> does that prior art deal with the project being concurrently built with different target directories?

I'm not fully understand the meaning, please correct me if I guessed wrong.

1. If I understand this question as "does package system build concurrently after creating actual build directories?" yes, all of these systems do build concurrently if possible. 
2. If I understand this question as "we have different CPU/build targets, how we would create a folder structure for them?", as far as I understood this RFC, that only base `/path/to/project/target` folder would be moved, and the contents would stay the same, as done for example by Xcode and by cargo (at the moment): they create base folder and create subfolders as needed. This system works already and I don't see a reason to change it, except of the actual placement for project/_target_ folder

Reply:
Give me a shout if you need documentation references for these systems

Reply:
Done in fac2d9c0852b8b043d3baf22f3fe5313de743072

Reply:
Can we make sure we also cover the prior art?

Reply:
I assume prior arts for forward links ? barring `targo`, I have not found any (I suspect I just don't know the domain terms used by other build systems)

- `bazel`: nothing I could find, instead they document their naming scheme for the build paths, allowing users to very easily build their own forward link if wanted
- `go build`: builds a single binary directly in the root (or cwd)
- `make`, `configure`, `cmake`: everyone does it differently but the convention is often `<project root>/build/` dir or sometimes a `dist/` one

I haven't reviewed every build system in existence, I could look toward the JS ecosystem, they probably have one package manager with forward link but I have never worked with JS so I wouldn't even know where to start

Reply:
Yeah, I'm not aware of any prior art for forward links for the target directory itself. I believe Buck2 creates symlinks (or maybe hard links) for executables to a content-addressed store.

Reply:
Was referring to the Ports system talked about in @eirnym's posts.

Reply:
I would be interested in links to docs for those then, because I know nothing about them and build tools are not something I'm willing to spend time diving in to find that docs don't even exist in the end.